### PR TITLE
Refactor test/Main.hs into separate test suites.

### LINKED ISSUE #510: Break down test/Prop.hs and test/Main.hs into smaller modules
#146 is about organizing tests by source modules. This issue is from build perspective. First of all we need to enable optimization options, same as for benchmarks or at least `-O2` for all tests. This is also required for `fusion-plugin` to work on tests. But `-O2` currently takes inordinate amount of time and memory (~6GB) to compile `Prop.hs` and `Main.hs`. We need to break down these modules into smaller modules so that we can compile them in reasonable amount of time and memory. At the same time the tests can go into their own respective modules corresponding to source modules where applicable.

See also #433 . The test modules can also be broken down on the same lines.

### PR DESCRIPTION
**Pull request by pranaysashank**
Closes #510

### CODE REVIEW

File: test/Streamly/Test/Internal/Prelude.hs [hunk 1]
```diff
@@ -88,9 +91,34 @@ unfold0 = monadicIO $ do
     a <- pick $ choose (0, max_length `div` 2)
     b <- pick $ choose (0, max_length)
     let unf = UF.supply (UF.enumerateFromToIntegral b) a
-    ls <- S.toList $ SI.unfold0 unf
+    ls <- S.toList $ IS.unfold0 unf
     return $ ls == [a..b]
 
+testFromCallback :: IO Int
+testFromCallback = do
```
Comment (pranaysashank) [inline test/Streamly/Test/Internal/Prelude.hs:98]: This function should instead go in the prelude test suite.

File: streamly.cabal [hunk 2]
```diff
@@ -597,77 +582,96 @@ test-suite pure-streams-streamly
     , hspec             >= 2.0   && < 3
   default-language: Haskell2010
 
+test-suite Prelude
+  import: test-options, prelude-test-options
+  type: exitcode-stdio-1.0
+  main-is: Streamly/Test/Prelude.hs
+  ghc-options: -main-is Streamly.Test.Prelude.main
```
Comment (adithyaov) [inline streamly.cabal:589]: Why do you need `-main-is`?

Comment (adithyaov) [reply streamly.cabal:589]: `-main-is` seems to be there is all the `test-suite`s

File: test/Streamly/Test/Prelude/Concurrent.hs [hunk 3]
```diff
@@ -60,6 +65,23 @@ mvarSequenceOp mv n x = do
     else dbgMVar ("put mvarSequenceOp" <> msg)
             (replicateM_ (fromIntegral n) (putMVar mv ())) >> return x
 
+checkCleanup :: IsStream t
+    => Int
+    -> (t IO Int -> SerialT IO Int)
+    -> (t IO Int -> t IO Int)
+    -> IO ()
+checkCleanup d t op = do
```
Comment (adithyaov) [inline test/Streamly/Test/Prelude/Concurrent.hs:73]: This is duplicated. It exists in `Fold` as well.

Comment (adithyaov) [reply test/Streamly/Test/Prelude/Concurrent.hs:73]: and `Parallal`

File: test/Streamly/Test/Prelude/WSerial.hs [hunk 4]
```diff
@@ -64,6 +79,29 @@ main = hspec
...
+            ]
+            wSerially
+        composeAndComposeSimpleWSerially
+            "WSerial <> "
+            [ [1, 4, 2, 7, 3, 5, 8, 6, 9]
+            , [1, 7, 4, 8, 2, 9, 5, 3, 6]
+            , [1, 4, 3, 7, 2, 6, 9, 5, 8]
+            , [1, 7, 4, 9, 3, 8, 6, 2, 5]
+            ]
+            wSerially
```
Comment (adithyaov) [inline test/Streamly/Test/Prelude/WSerial.hs:103]: Did you miss `composeAndComposeSimpleWSerially` `serially`

File: streamly.cabal (hunk 2 above)
Comment (pranaysashank) [reply streamly.cabal:589]: It allows me to use a module name other than Main, which makes it easier to compile all the test files with ghc directly.

File: test/Streamly/Test/Prelude/WSerial.hs (hunk 4 above)
Comment (pranaysashank) [reply test/Streamly/Test/Prelude/WSerial.hs:103]: It's in Prelude.Serial

File: test/Streamly/Test/Prelude/Concurrent.hs (hunk 3 above)
Comment (pranaysashank) [reply test/Streamly/Test/Prelude/Concurrent.hs:73]: Should be refactored into the Prelude module

**Review by adithyaov — APPROVED** [review]
LGTM.

**Comment by harendra-kumar** [comment]
Why are the concurrency tests failing in the appveyor? Is it intermittent? They were passing earlier I suppose.

**Review by harendra-kumar — APPROVED** [review]

**Comment by pranaysashank** [comment]
Looks like an intermittent failure, it wasn't present for the previous commit.

**Comment by pranaysashank** [comment]
Perhaps the delay in the test isn't sufficient

**Pull request merged**